appinventor/blocklyeditor/src/replmgr.js

```diff
@@ -1683,37 +1813,43 @@ Blockly.ReplMgr.putAsset = function(projectid, filename, blob, success, fail, fo
 
     var conn = goog.net.XmlHttp();
     var arraybuf = new ArrayBuffer(blob.length);
+    var arrayview = new Uint8Array(arraybuf);
+    for (var i = 0; i < blob.length; i++) {
+        arrayview[i] = blob[i];
+    }
     var rs = top.ReplState;
     var encoder = new goog.Uri.QueryData();
     //var z = filename.split('/'); // Remove any directory components
     //encoder.add('filename', z[z.length-1]); // remove directory structure
     var z = filename.slice(filename.indexOf('/') + 1, filename.length); // remove the asset directory
     encoder.add('filename', z); // keep directory structure
 
-    conn.retries = 3;
-    conn.open('PUT', rs.baseurl + '?' + encoder.toString(), true);
-    conn.onreadystatechange = function() {
-        if (this.readyState == 4 && this.status == 200) {
-            if (success) {      // process callbacks
-                success();
-            }
-        } else if (this.readyState == 4) {
-            if (this.retries > 0) {
-                this.retries--;
-                this.open('PUT', rs.baseurl + '?' + encoder.toString(), true);
-                this.send(arraybuf);
-            }
-            if (fail) {
-                fail();
-            }
-        }
-    };
 
-    var arrayview = new Uint8Array(arraybuf);
-    for (var i = 0; i < blob.length; i++) {
-        arrayview[i] = blob[i];
+    if (rs.proxy) {
+        rs.proxy.postMessage(['asset', encoder.toString(), arraybuf], rs.proxy_origin);
+        success();              // What happens if we fail?
```

Comment:
Maybe document what happens if we fail.

Reply:
Actually, I'm not sure what happens. postMessage() isn't documented to return an error, though other sources reference that it can throw an exception. However, I don't think it ever will given the way we use it (no use of shared memory).